[lldb] Make logging machinery type-safe

medismailben opened the pull request:
This patch makes use of c++ type checking and scoped enums to make
logging statements shorter and harder to misuse.

Defines like LIBLLDB_LOG_PROCESS are replaces with LLDBLog::Process.
Because it now carries type information we do not need to worry about
matching a specific enum value with the right getter function -- the
compiler will now do that for us.

The main entry point for the logging machinery becomes the GetLog
(template) function, which will obtain the correct Log object based on
the enum type. It achieves this through another template function
(LogChannelFor<T>), which must be specialized for each type, and should
return the appropriate channel object.

This patch also removes the ability to log a message if multiple
categories are enabled simultaneously as it was unused and confusing.

This patch does not actually remove any of the existing interfaces. The
defines and log retrieval functions are left around as wrappers around
the new interfaces. They will be removed in follow-up patch.

Differential Revision: https://reviews.llvm.org/D117490

labath committed:
[lldb/linux] Fix a bug in wait status handling

The MonitorCallback function was assuming that the "exited" argument is
set whenever a thread exits, but the caller was only setting that flag
for the main thread.

This patch deletes the argument altogether, and lets MonitorCallback
compute what it needs itself.

This is almost NFC, since previously we would end up in the
"GetSignalInfo failed for unknown reasons" branch, which was doing the
same thing -- forgetting about the thread.

labath committed:
[lldb-server/linux] Fix waitpid for multithreaded forks

The lldb-server code is currently set up in a way that each
NativeProcess instance does its own waitpid handling. This works fine
for BSDs, where the code can do a waitpid(process_id), and get
information for all threads in that process.

The situation is trickier on linux, because waitpid(pid) will only
return information for the main thread of the process (one whose tid ==
pid). For this reason the linux code does a waitpid(-1), to get
information for all threads. This was fine while we were supporting just
a single process, but becomes a problem when we have multiple processes
as they end up stealing each others events.

There are two possible solutions to this problem:
- call waitpid(-1) centrally, and then dispatch the events to the
  appropriate process
- have each process call waitpid(tid) for all the threads it manages

This patch implements the second approach. Besides fitting better into
the existing design, it also has the added benefit of ensuring
predictable ordering for thread/process creation events (which come in
pairs -- one for the parent and one for the child). The first approach
OTOH, would make this ordering even more complicated since we would
have to keep the half-threads hanging in mid-air until we find the
process we should attach them to.

The downside to this approach is an increased number of syscalls (one
waitpid for each thread), but I think we're pretty far from optimizing
things like this, and so the cleanliness of the design is worth it.

The included test reproduces the circumstances which should demonstrate
the bug (which manifests as a hung test), but I have not been able to
get it to fail. The only place I've seen this failure modes are very
rare hangs in the thread sanitizer tests (tsan forks an addr2line
process to produce its error messages).

Differential Revision: https://reviews.llvm.org/D116372

labath committed:
[lldb] Make logging machinery type-safe

This patch makes use of c++ type checking and scoped enums to make
logging statements shorter and harder to misuse.

Defines like LIBLLDB_LOG_PROCESS are replaces with LLDBLog::Process.
Because it now carries type information we do not need to worry about
matching a specific enum value with the right getter function -- the
compiler will now do that for us.

The main entry point for the logging machinery becomes the GetLog
(template) function, which will obtain the correct Log object based on
the enum type. It achieves this through another template function
(LogChannelFor<T>), which must be specialized for each type, and should
return the appropriate channel object.

This patch also removes the ability to log a message if multiple
categories are enabled simultaneously as it was unused and confusing.

This patch does not actually remove any of the existing interfaces. The
defines and log retrieval functions are left around as wrappers around
the new interfaces. They will be removed in follow-up patch.

Differential Revision: https://reviews.llvm.org/D117490

labath committed:
[lldb] Fix mac build for D117490

This is exactly that kind of a API misuse that the patch was meant to
detect.

labath committed:
[lldb] Fix windows build for D117490

I forgot to update ProcessWindowsLog to the new API.

labath committed:
[lldb] Fix ProcessKDPLog for the logging refactor

labath committed:
[lldb] Convert POSIXLog to use the new API

labath committed:
[lldb] Convert ProcessWindowsLog to the new API

labath committed:
[lldb] Convert DWARF log to the new API

This also deletes some dead log statements (log initialization commented
out).

labath committed:
[lldb] Convert ProcessGDBRemoteLog to the new API

labath committed:
[lldb] Convert Process KDP Log to the new API

labath committed:
[lldb] Fix windows&mac builds for c34698a811b13

medismailben committed:
[lldb/swift] Adopt new type-safe logging in SwiftLog (NFC)

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

labath committed:
[lldb] Convert "LLDB" log channel to the new API

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

labath committed:
[lldb] Rename Logging.h to LLDBLog.h and clean up includes

Most of our code was including Log.h even though that is not where the
"lldb" log channel is defined (Log.h defines the generic logging
infrastructure). This worked because Log.h included Logging.h, even
though it should.

After the recent refactor, it became impossible the two files include
each other in this direction (the opposite inclusion is needed), so this
patch removes the workaround that was put in place and cleans up all
files to include the right thing. It also renames the file to LLDBLog to
better reflect its purpose.

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben committed:
[lldb] Reflect Logging.h -> LLDBLog.h renaming in Swift Plugins

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
This depends on #3958

medismailben commented:
@swift-ci test

kastiglione reviewed (approved):
thanks for picking this!

kastiglione commented on lldb/source/Plugins/Language/Swift/LogChannelSwift.cpp:
```diff
@@ -0,0 +1,44 @@
+  return g_channel;
+}
+
+void LogChannelSwift::Initialize() {
```
I don't see where this `Initialize()` is called.

medismailben replied:
Resolved in 0f528214795aa15b4cc250ba41ce271ca6af2e70

medismailben committed:
[lldb/Swift] Initialize the LogChannelSwift logging channel

This initializes the swift logging channel in the Swift Language Plugin
initializer.

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben committed:
[lldb/Swift] Separate Swift logging from core logging (NFC)

This patch removes the Swift logging channel from lldb's core logging file.
It moves it into a separate file in the Swift Language Plugin, so it can be
build conditionally when Swift support is enabled.

This also applies to the the `swift-healthcheck` command object, which
will now only be available when Swift support is enabled.

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test

JDevlieghere commented:
Thanks Ismail!

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test linux

medismailben commented:
#3906 introduces a new merge conflict 🥲 @swift-ci test again

medismailben commented:
@swift-ci test and merge

medismailben commented:
@swift-ci test

medismailben committed:
[lldb] Convert Swift-LLDB to type-safe logging API

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

compnerd commented:
This caused a regression in the build; you added a new source file but didn't correctly update the build system to indicate that it is an optional source file, which breaks the LLVM build.

medismailben commented:
Hi @compnerd, I'm just seeing this. Sorry I missed that and thanks for fixing it!
